packages/core/src/browser/shell/application-shell.ts

```diff
@@ -638,6 +645,14 @@ export class ApplicationShell extends Widget {
         if (mainPanel) {
             this.mainPanel.restoreLayout(mainPanel);
             this.registerWithFocusTracker(mainPanel.main);
+            const widgets = toArray(this.mainPanel.widgets());
+            if (mainPanelPinned && mainPanelPinned.length === widgets.length) {
```

Comment:
See above for a way to handle the uncertainty of whether the widgets restored will be the same as the widgets whose state was saved.